### DISCUSSION TITLE
Understanding how to use the `GravityCompensationHelper` class in iDynTree

### DISCUSSION THREAD

**Discussion by nicholaspalomo** [robotology/community#547] [Q&A] [answered]
Hello and thank you very much for your work on the iDynTree library.

I'm currently using it in a project, and I would like to have a simple joint PD controller + gravity compensation torques running on my robot in simulation (Ignition Gazebo).

My question pertains to proper usage of the [GravityCompensationHelper](https://robotology.github.io/idyntree/master/classiDynTree_1_1GravityCompensationHelper.html#a43e80215c956c765e241b5165ad6ee37) class in the iDynTree library.

From my understanding of the documentation and the examples on the official iDynTree repository, I need to do the following in order to get the gravity compensation torques for my robot.

1. Create an `iDynTree::ModelLoader` object:
```
iDynTree::ModelLoader modelLoader;
```

2. Load a URDF of my robot into the `ModelLoader` object:
```
auto success = modelLoader.loadModelFromFile(urdfPath);
```

3. Load the model into my `GravityCompensationHelper` object:
```
iDynTree::GravityCompensationHelper gravComp;
gravComp.loadModel(modelLoader.model(), baseFrame);
```
where `baseFrame` here is a string containing the name of the base frame in my URDF model (e.g. `base_link`).

4. Then, I need to update the kinematics from the gravity vector:
```
iDynTree::JointPosDoubleArray jointPos(7); // in my case, my robot has 7 joints. Not shown here, I populate this array with sensible values.
iDynTree::FrameIndex baseFrameIndex(0); // in my case, my robot is fixed-base with the base frame being the 0th one in my URDF

iDynTree::Vector3 gravity; // I define a gravity vector in the base frame
            gravity.zero();
            gravity[2] = -9.80665;

gravComp.updateKinematicsFromGravity(jointPos, baseFrameIndex, gravity);

iDynTree::JointDOFsDoubleArray jointTorques;
bool success = gravComp.getGravityCompensationTorques(jointTorques);
```
If all went well, `jointTorques` should now contain the gravity compensation torques, g(q), in the dynamics equation for my robot that I can then use as a feedforward term in my PD controller.

Is this correct? Or am I missing a step somewhere in here?

Thanks!!

**Comment by traversaro** [comment, marked as answer]
Everything seems to be correct at a first glance. Do you obtain strange or unexpected results by using the library like that?

**Reply by nicholaspalomo** [reply]
Hi, @traversaro. Sorry for my late response. I think I got it. Thank you very much for your answer. I ended up doing the following, which works well for my application:

I set my joint torques as:
```
error = targetJointPositions - robot_->generalizedCoordinates().tail(numJoints_);
jointTorques = P_ * error - D_ * robot_->generalizedVelocities().tail(numJoints_);
jointTorques = freeFloatingMassMatrix.bottomRightCorner(7, 7) * jointTorques.segment<7>(0)) + biasTorques;
```
where `freeFloatingMassMatrix` is returned by a method I wrote:
```
        Eigen::MatrixXd getFreeFloatingMassMatrix() {

            iDynTree::MatrixDynSize M(numJointsIk()+6, numJointsIk()+6);
            fk_->getFreeFloatingMassMatrix(M);

            return iDynTree::toEigen(M);

        }
```
and `fk_` is a pointer to an object of type `iDynTree::KinDynComputations`.
Similarly, I get the `biasTorques` from a function:
```
        Eigen::VectorXd getBiasTorques() {
            
            fk_->generalizedBiasForces(*genBiasForce_);

            iDynTree::JointDOFsDoubleArray& jointTorques = genBiasForce_->jointTorques();

            return Eigen::Map<Eigen::VectorXd>(jointTorques.data(), numJointsIk(), 1);

```
where `genBiasForce_` is a pointer to an object of type `iDynTree::FreeFloatingGeneralizedTorques`.

(credit: I stole this idea from your code here -> https://github.com/robotology/gym-ignition/blob/devel/scenario/src/controllers/src/ComputedTorqueFixedBase.cpp#L205)

Thanks again!
